### PR TITLE
Ros2 port traffic light map

### PR DESCRIPTION
**Pull request by simon-t4**
Porting traffic light map detector module to ROS2.

Module has standard ROS interfaces, port:
[x] CMakeLists to use ament
[x] convert messages, rclcpp calls
[x] setup subscribers and publishers

### CODE REVIEW

**Comment by simon-t4** [comment]
@mitsudome-r One issue is that roi vibration parameters need a default value. I have assigned 0.0 default values but I ma not sure this is appropriate.

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp [hunk 1]
```diff
@@ -74,11 +88,12 @@ void MapBasedDetector::cameraInfoCallback(const sensor_msgs::CameraInfo::ConstPt
     camera_pose_stamped.pose.orientation.z = transform.transform.rotation.z;
     camera_pose_stamped.pose.orientation.w = transform.transform.rotation.w;
   } catch (tf2::TransformException & ex) {
-    ROS_WARN_THROTTLE(5, "cannot get transform frome map frame to camera frame");
+    RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 5,
```
Comment (mitsudome-r) [inline perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp:91]: It seems the units for THROTTLE has changed to ms from seconds so please change this to `5000`

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp [hunk 2]
```diff
@@ -375,15 +399,15 @@ void MapBasedDetector::publishVisibleTrafficLights(
     point.z = tf_camera2tl.getOrigin().z();
     marker.points.push_back(point);
 
-    marker.lifetime = ros::Duration(0.2);
+    marker.lifetime = rclcpp::Duration(0.2);
```
Comment (mitsudome-r) [inline perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp:402]: I think this is implicitly calling constructor with nanosec.
You can use `rclcppp::Duration::from_seconds(0.2)` function instead.

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp [hunk 3]
```diff
@@ -17,54 +17,68 @@
...
+  autoware_perception_msgs::msg::TrafficLightRoiArray output_msg;
   output_msg.header = camera_info_ptr_->header;
-  geometry_msgs::PoseStamped camera_pose_stamped;
+  geometry_msgs::msg::PoseStamped camera_pose_stamped;
   try {
-    geometry_msgs::TransformStamped transform;
+    geometry_msgs::msg::TransformStamped transform;
     transform = tf_buffer_.lookupTransform(
-      "map", input_msg->header.frame_id, input_msg->header.stamp, ros::Duration(0.2));
+       "map", input_msg->header.frame_id, tf2::TimePointZero);
```
Comment (mitsudome-r) [inline perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp:81]: I think we've found out that lookupTransform with timeout works if timestamp is not zero so could you change this to use the exact time with timeout?
e.g. 
```
tf2_buffer_.lookupTransform(target_frame, source_frame, time_stamp, rclcpp::Duration::from_seconds(0.2));
```

File: perception/traffic_light_recognition/traffic_light_map_based_detector/CMakeLists.txt [hunk 4]
```diff
@@ -1,69 +1,38 @@
...
 ###########
 ## Build ##
 ###########
 
 include_directories(
-  include
-  ${catkin_INCLUDE_DIRS}
   ${EIGEN3_INCLUDE_DIR}
+  include
+  src/autoware/pilot.auto/map/lanelet2_extension/include
```
Comment (mitsudome-r) [inline perception/traffic_light_recognition/traffic_light_map_based_detector/CMakeLists.txt:23]: These two lines are not necessary.

**Review by mitsudome-r — COMMENTED** [review]
It looks good overall with just a few comments.

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp (hunk 3 above)
Comment (simon-t4) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp:81]: Do we have a working example? BufferCore doesn't seem to accept such a function. Should we switch to back Buffer?

File: perception/traffic_light_recognition/traffic_light_map_based_detector/CMakeLists.txt (hunk 4 above)
Comment (simon-t4) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/CMakeLists.txt:23]: Thanks, removed

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp (hunk 2 above)
Comment (simon-t4) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp:402]: Thanks, done.

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp (hunk 1 above)
Comment (simon-t4) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp:91]: Thanks, done.

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp (hunk 3 above)
Comment (simon-t4) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp]: @mitsudome-r I have pushed the above changes, excluding the lookuptransform. Let me know if you want to to revisit this.

File: perception/traffic_light_recognition/traffic_light_map_based_detector/launch/traffic_light_map_based_detector.launch.xml [hunk 5]
```diff
@@ -7,11 +7,10 @@
...
-    <remap from="~input/vector_map" to="$(arg input/vector_map)" />
-    <remap from="~input/camera_info" to="$(arg input/camera_info)" />
-    <remap from="~input/route" to="$(arg input/route)" />
-    <remap from="~output/rois" to="$(arg output/rois)" />
-    <rosparam command="load" file="$(find traffic_light_map_based_detector)/config/traffic_light_map_based_detector.yaml" />
+  <node pkg="traffic_light_map_based_detector" exec="traffic_light_map_based_detector_node" name="traffic_light_map_based_detector" output="screen">
+    <remap from="/input/vector_map" to="$(val input/vector_map)" />
+    <remap from="/input/camera_info" to="$(val input/camera_info)" />
+    <remap from="/input/route" to="$(val input/route)" />
+    <remap from="/output/rois" to="$(val output/rois)" />
```
Comment (mitsudome-r) [inline perception/traffic_light_recognition/traffic_light_map_based_detector/launch/traffic_light_map_based_detector.launch.xml:14]: `val` must be `var`

**Review by mitsudome-r — COMMENTED** [review]
I noticed one more small fix in launch file.

**Comment by simon-t4** [comment]
@mitsudome-r Done!

File: perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp (hunk 3 above)
Comment (cvasfi) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp]: > I think we've found out that lookupTransform with timeout works if timestamp is not zero so could you change this to use the exact time with timeout?
@mitsudome-r  Strangely that is only the case with `waitForTransform` I think, reading the logic of `tf2_ros::lookupTransform`, it seems like it should work.

I am still a bit confused with what's the issue here so here's my take:

* If the aim is to have `lookupTransform` wait until the transform is available, then using `tf2_ros::Buffer::lookupTransform` with the message timestamp + timeout should be used.
* If the aim is to work with the current state of the buffer and not wait for anything then:
   * Use lookupTransform with exact time stamp -> You might not succeed if the data has not arrived just yet.
   * Or do something more fail-safe:

```cpp
if(tf_buffer_.can_transform(/* */, time_stamp)){
   tf = tf_buffer_. lookupTransform(/* */, time_stamp);
} else {  // Data hasn't arrived yet, use the best we got:
   tf = tf_buffer_. lookupTransform(/* */, tf2::TimePointZero);
}
```

Comment (simon-t4) [reply perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp]: @cvasfi @mitsudome-r 
Thanks for the suggestions.
Used `tf2_ros::Buffer` and `lookupTransform` with message timestamp and timeout.

Something not obvious (to me), I had to instantiate the Buffer object with the node clock or else there is a compilation error in the TransformListener instantiation.

**Review by mitsudome-r — APPROVED** [review]

**Pull request merged**